src/box/txn_limbo.c

```diff
@@ -37,9 +37,75 @@
 #include "raft.h"
 #include "tt_static.h"
 #include "session.h"
+#include "trivia/config.h"
 
 struct txn_limbo txn_limbo;
 
+static int
+txn_limbo_write_synchro(struct txn_limbo *limbo, uint16_t type, int64_t lsn,
+			uint64_t term, struct vclock *vclock);
+
+static void
+txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn);
+
+/**
+ * Write a confirmation entry to the WAL. After it's written all the
+ * transactions waiting for confirmation may be finished.
+ */
+static int
+txn_limbo_write_confirm(struct txn_limbo *limbo, int64_t lsn)
+{
+	assert(lsn > limbo->confirmed_lsn);
+	assert(!limbo->is_in_rollback);
+	return txn_limbo_write_synchro(limbo, IPROTO_RAFT_CONFIRM, lsn, 0,
+				       NULL);
+}
+
+static int
+txn_limbo_worker_bump_confirmed_lsn(struct txn_limbo *limbo)
+{
+	assert(limbo->is_in_rollback ||
+	       limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
```

Comment:
How is the LSN comparison related to `is_in_rollback`? How does rollback being set allow `volatile_confirmed_lsn` to become `< confirmed_lsn `?

I applied this diff and the replication-luatest/ + box-luatest/ tests passed. Did I miss something here?
```Diff
diff --git a/src/box/txn_limbo.c b/src/box/txn_limbo.c
index e1a01ec6c8..ce41b35c1a 100644
--- a/src/box/txn_limbo.c
+++ b/src/box/txn_limbo.c
@@ -64,8 +64,7 @@ txn_limbo_write_confirm(struct txn_limbo *limbo, int64_t lsn)
 static int
 txn_limbo_worker_bump_confirmed_lsn(struct txn_limbo *limbo)
 {
-	assert(limbo->is_in_rollback ||
-	       limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
+	assert(limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
 	while (txn_limbo_is_owned_by_current_instance(limbo) &&
 	       limbo->volatile_confirmed_lsn > limbo->confirmed_lsn) {
 		if (limbo->is_in_rollback)
@@ -80,8 +79,7 @@ txn_limbo_worker_bump_confirmed_lsn(struct txn_limbo *limbo)
 		ERROR_INJECT_YIELD(ERRINJ_TXN_LIMBO_WORKER_DELAY);
 		txn_limbo_read_confirm(limbo, volatile_confirmed_lsn);
 	}
-	assert(limbo->is_in_rollback ||
-	       limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
+	assert(limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
 	return 0;
 }
 ```

Reply:
I originally thought that the `volatile_confirmed_lsn` could possibly go lower than `confirmed_lsn` during the promote/demote request execution, but after our discussion it seems like the confirmed lsns can only monotonically grow, so the invariant `volatile_confirmed_lsn >= confirmed_lsn` always holds.

Applied your diff.